src/lsm/manifest_log.zig

```diff
@@ -684,26 +743,20 @@ pub fn ManifestLogType(comptime Storage: type) type {
             }
 
             log.debug("compacted: checksum={} address={} frees={}/{}", .{
-                block_reference.checksum,
-                block_reference.address,
+                head_checksum,
+                head_address,
                 frees,
                 block_schema.entry_count,
             });
 
-            // Blocks may be compacted if they contain frees, or are not completely full.
-            // (A partial block may be flushed as part of a checkpoint.)
-            assert(frees > 0 or block_schema.entry_count < schema.Manifest.entry_count_max);
-            // At most one block could have been filled by the compaction.
+            // Blocks are compacted in sequence – not skipped, even if no entries will be freed.
+            // (That should be rare though, since blocks are large.)
+            // This is necessary to update the block's "previous block" pointer in the header.
+            maybe(frees == 0);
             assert(manifest_log.blocks_closed <= 1);
```

Comment:
We need enough blocks to accommodate all of the appends by table compaction (inserts, update snapshot_max, etc).
The number of blocks this can fill (in the worst case, where we compact all trees+levels) is dependent on how many entries can fit in each block -- i.e., it depends on the block size.
For example, in the production config, with 128KiB blocks, we need as many as 3 blocks.

In addition, right now we always compact exactly one block per half-bar. The next manifest PR will change that, to speed up compaction as the manifest size increases.

Reply:
>We need enough blocks to accommodate all of the appends by table compaction (inserts, update snapshot_max, etc).

But why we need it? Can't we spill blocks to disk as they are being filled, instead of writing them in one go in the end?

Reply:
One constraint here is that we need to make sure that all of the inserts/updates/etc are appended in a deterministic order, regardless of the order that the table compactions actually finish in.

The way we handle that right now is: we wait until all of the (half-bar's) table compactions finish, and then we synchronously apply all the compactions' appends at once.